docs2/site/docs/getting-started/custom-scalars.md

```diff
@@ -183,105 +213,303 @@ public class ExampleSchema : Schema
+    if (value == null)
+        return new NullValue();
+
+    if (value is Vector3 vector3)
+    {
+        return new StringValue($"{vector3.X},{vector3.Y},{vector3.Z}");
+    }
 
-    return value is StringValue stringValue
-        ? ParseValue(stringValue.Value)
-        : null;
+    return ThrowASTConversionError(value);
 }
+```
+
+With these changes, a variable can be parsed as a string or as a structured object, and is always returned as a structured object.
+
+## ValueConverter
+
+When `GetArgument<T>` is called, the argument value is coerced to the requested type via the `ValueConverter`.
+No conversion takes place when the requested type matches the type of the object or scalar (the type returned from
+`ParseLiteral` or `ParseValue`). But you can also use the value converter to assist with input deserialization.
+
+For instance, you may be using `IdGraphType` within your schema as unique identifiers for your data objects. Pursuant
+to the GraphQL specification, these identifiers should be passed as strings such as in the below example:
 
+```graphql
+{
+    widget (id: "3") {
+        name
+    }
+}
 ```
 
-This is necessary since the query executor converts all arguments and variables to `IValue` instances before coercing them to their server-side representation using `ParseLiteral`. `Vector3` instances parsed from variables will be converted to the more specific `Vector3Value` type.
+However, your code may use integer identifiers in the data access layer. So when you call `context.GetArgument<int>("id")`,
+GraphQL.NET calls the value converter to convert the string to an integer.
 
-7. Implement `ScalarGraphType.Serialize` so `Vector3` instances can be sent to the client.
+The value converter can be extended globally by calling the static method `Register` as follows:
 
 ```csharp
-// In Vector3Type
+ValueConverter.Register<Vector3, string>(v => $"{v.X},{v.Y},{v.Z}");
+```
 
-public override object Serialize(object value)
+The above method registers a conversion from the `Vector3` format to a `string`. Since the registration is static,
+it should only be done once per application lifetime. For instance, in a static constructor of your schema.
+
+```csharp
+public class MySchema : Schema
 {
-    return ValueConverter.ConvertTo(value, typeof(Vector3));
+    static MySchema()
+    {
+        ValueConverter.Register<Vector3, string>(v => $"{v.X},{v.Y},{v.Z}");
+    }
+
+    ...
 }
 ```
 
-This implementation may surprise you. Why is `Serialize`, which is used for output, implemented identically to `ParseValue`, which is used for input? Why does `Serialize` return an object, rather than a string or byte array? It helps to understand a few internals of the library.
+## Null values
 
-- `Serialize` will be called during query execution, and should be passed an instance of `Vector3` from a field resolver.
+Custom scalars process and handle null values during serialization and deserialization. This allows
```

Comment:
> This allows
for custom scalars that can assist when you have database values such as 0 that should represent null
when exposed outside the schema.

It is difficult for me to understand. There is no mistake here?

Reply:
I have a database where that is the case.  Another scenario I run into often is null vs empty strings.  Specifically, some databases hold null where they should represent empty strings, or some databases that hold empty strings where they should represent null.

If you have a suggestion of how I can explain this better, or a better sample to show altogether, I'm listening.

Reply:
I'm talking about a long sentence to read and understand to me as not a native speaker. Maybe split to shorter ones?

Reply:
Two that, two when.

Reply:
See suggestion here: https://github.com/graphql-dotnet/graphql-dotnet/pull/2357/files#r593837457

Modify or merge as desired.

Reply:
done